### PR TITLE
Adding fine-tuning

### PR DESCRIPTION
**Pull request by daanelson**
This PR updates this script to enable lora fine-tuning of llama models. Changes:

- switched tuning to bf16
- added in gradient checkpointing
- included `max_length` for the tokenizer to truncate training inputs to the 4k context window
- Updated relevant packages + python version

### CODE REVIEW

**Comment by rajko-rad** [comment]
Thank you @daanelson this is awesome!!
